miniflux/v2
# Updating read/star icons to SVGs

Initial change to start the discussion. Here is the overview of the changes I did to address https://github.com/miniflux/v2/issues/918:

0) template/common.go, ui/static/js.go are generated files (nothing more than a result of 'go generate'). Since they're tracked under source control already, I've included latest versions with my change, let me know if I should exclude them from the change.

The other three files have been changed to implement SVG icons:

1) template/html/common/icons.html: I've added SVG "unstar" icon, since it was missing in the set. It's a copy of a star icon with a fill added. This is different from all other icons, since none of them has any presentational styles on the geometry itself, however I can't think of a way to implement a "filled star" without adding fill on the path. It's possible to add a CSS style, though personally I wouldn't recommend that - it'll have the same effect, but would be potentially harder to understand and debug in the future.

2) template/html/common/item_meta.html: this file contains two principal changes:

- Prior to the change emoji symbol was part of the label (within <span class="icon-label"> element), in order to replace this symbol with SVG I've moved SVG out of this element consistently with other SVG + label items.
- Prior to the change data-label-read/unread/star/unstar contained both the label and the text. In order to implement script based toggle I've split attributes to two groups. Existing data-label-* now contain label only (e.g. "Read"/"Unread") and I've added data-icon-* attributes that contain the SVG icons. Note the single quote required to escape double quotes in SVG (more on that later in follow-up ideas).

3) ui/static/js/app.js: this is trivial change after change 2 above. Prior to the change both the symbol and text were replaced inside the <span class="icon-label"> element in one operation from the data-label-* attributes. With the split - I'm adding icon followed by label.

Testing done:
- Verified visually and with the DevTools that new icons are used in Unread, Starred, History, Feeds (all feed entries)
- Verified that in all sections the toggle replacement happens correctly

Notes and ideas for follow-ups:

- I've noticed that template/html/entry.html has emoji symbols for read/unread, star/unstar. I couldn't determine how this template is used and don't see emoji symbols on my local setup with my change. Let me know if I should update this file and what's the best way to test changes to it.

- My current change tries to be consistent with the existing implementation, and is a bit more verbose as a result. To implement it in a more clean way I can suggest few alternatives:
  - Add both icons, hide one based on state. In the toggle functions change the text, toggle visibility on the icons
  - Move icons to styles and use as background images. In the toggle functions change the text and class to use different icons. Would be fairly intrusive change.
  - Change from storing text and icons on every element and add <template> elements to the page. In the toggle script replace icon + label contents with template contents. 

Happy to brainstorm more and address any other feedback on the change!
